File: leetcode/algorithms/stack/min-stack.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Min Stack"
*/

    //stack + variable approach

class MinStack {

    /** initialize your data structure here. */
    Stack<Integer> stack = null;
    Integer min = null;
    
    public MinStack() {
        stack = new Stack<Integer>();
        min = null;
        
    }
    
    public void push(int x) {
        if(min==null){
            min=x;
        } else if(min>=x){
            stack.push(min);
            min = x;
        }
        stack.push(x);
    }
    
    public void pop() {
        Integer poppedInt = stack.pop();
        if(poppedInt.intValue()==min.intValue()){
            if(stack.size()>0){
                min = stack.pop();
            } else {
                min=null;
            }
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: leetcode/algorithms/stack/min-stack.ts
/*
    leetcode
    category: "Algorithms"
    problem title: "Min Stack"
*/


class MinStack {
    _stack: number[];
    min: number | null;
    constructor() {
        
        this._stack = [];
        this.min = null;

    }

    push(x: number): void {
        
        if(this.min===null){
            this.min = x;
        } else if (x<=this.min){
            this._stack.push(this.min);
            this.min = x;
        }
        this._stack.push(x);

    }

    pop(): void {
        const poppedVal: number | null = this._stack.pop();

        if(poppedVal===this.min){
            this.min = this._stack.length===0 ? null : this._stack.pop();
        }
        

    }

    top(): number {
        
        return this._stack[this._stack.length-1];

    }

    getMin(): number {
        
        return this.min;

    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */